### PR TITLE
chore: compare xml parsers

### PR DESCRIPTION
**Pull request by justinemmanuelmercado**

### CODE REVIEW

**Comment by justinemmanuelmercado** [comment]
XML Parsers to benchmark

1. `fast-xml-parser`
2. `xml2js`

**Comment by justinemmanuelmercado** [comment]
| Name         | Run 1   | Run 2  | Run 3 |
| ------------ | ------ | -------- | -------- |
| `fast-xml-parser` |  |     |       |

**Comment by justinemmanuelmercado** [comment]
From `xml2js`
Screen Shot 2020-07-22 at 6 54 00 PM

This is cool, and I definitely didn't see this earlier.

In `GetCompetitivePricingForASINResponse`'s fixture https://github.com/ScaleLeap/amazon-mws-api-sdk/blob/de12f95866581d34fa9e9cd485a64fd2a39a3182/test/unit/__fixtures__/products_get_competitive_pricing_for_asin.xml#L32-L36

`fast-xml-parser` just gets the value and disregards the `condition` attribute which makes 

```
OfferListingCount` = `[11,1,10]
``` 
which isn't particularly useful 

but with `xml2js` 
```
OfferListingCount = [
{
  "$": {
    "condition": "Any",
  },
  "_": 11,
},
{
  "$": {
    "condition": "Used",
  },
  "_": 1,
},
{
  "$": {
    "condition": "New",
  },
  "_": 10,
}]
```

**Comment by moltar** [comment]
> where $ contains the attributes and _ contains the values, also $ and _ could be renamed to whatever variable in the settings

oh ya, that's pretty sweet!

**Comment by justinemmanuelmercado** [comment]
Changes required to make `xml2js` to work

1. setup
```typescript
await xml2js.parseStringPromise(response.data, {
        explicitArray: false,
        valueProcessors: [parseNumbers, parseBooleans],
        emptyTag: '',
        trim: true,
      })
```
2. Had to fix some fixtures that had invalid XML files

3. Some XML values i.e. https://github.com/ScaleLeap/amazon-mws-api-sdk/blob/de12f95866581d34fa9e9cd485a64fd2a39a3182/test/unit/__fixtures__/subscriptions_deregister_destination.xml#L2-L3

get parsed to `"/n   "` instead of `''` see the following difference in `Subscriptions` unit test snapshot https://github.com/ScaleLeap/amazon-mws-api-sdk/pull/142/commits/846202fac3e3e5c97b13ba1aee10210eae72320d#diff-973ce94a6aa7a940f8bc0588b0401817

4. Parsing error message had to be changed https://github.com/ScaleLeap/amazon-mws-api-sdk/pull/142/commits
/846202fac3e3e5c97b13ba1aee10210eae72320d#diff-614f14483603a8f5f71537a5bec76491 
Probably not that big of change mostly pedantic

**Comment by justinemmanuelmercado** [comment]
I'm really liking `xml2js` over `fast-xml-parser` due to configurability and it's much more precise with its parsing (compare the snapshots of the old to the new in the latest commit). Also it validates the XML (it throws an error when the received XML is invalid, apparently`fast-xml-parser` is just silent). Will explore other options

File: test/unit/__snapshots__/finances.test.ts.snap [hunk 1]
```diff
@@ -1650,7 +1650,7 @@ Array [
                 "CurrencyAmount": 3.38,
                 "CurrencyCode": "USD",
               },
-              "ProductDescription": "Shacke Metal Shoe Horn 7.5&quot; inches - Double Sided Stainless
+              "ProductDescription": "Shacke Metal Shoe Horn 7.5\\" inches - Double Sided Stainless
```
Comment (moltar) [inline test/unit/__snapshots__/finances.test.ts.snap:1653]: That's weird. I think both parsers are wrong here. Original one retained HTML symbols. The new one just replaces it with \\ 

The only explanation I can think of is that maybe HTML escaped symbols aren't valid in xml?

File: test/unit/__snapshots__/subscriptions.test.ts.snap [hunk 2]
```diff
@@ -28,7 +28,8 @@ Array [
 
 exports[`sellers deregisterDestination returns the standard response if deregistration is succesful 1`] = `
 Array [
-  "",
+  "
```
Comment (moltar) [inline test/unit/__snapshots__/subscriptions.test.ts.snap:31]: Why do you think that is? Is it that Amazon returning a new line?

Comment (justinemmanuelmercado) [reply test/unit/__snapshots__/subscriptions.test.ts.snap:31]: I checked and it's due to the difference of how this
```xml
   <DeregisterDestinationResult>
    </DeregisterDestinationResult>
```
and this
```xml
<DeregisterDestinationResult></DeregisterDestinationResult>
```
or this
```xml
<DeregisterDestinationResult/>
```
gets parsed

Comment (justinemmanuelmercado) [reply test/unit/__snapshots__/subscriptions.test.ts.snap:31]: The first one gets parsed to `"\n    "` and the other two get parsed to `""`

File: test/unit/__snapshots__/finances.test.ts.snap (hunk 1 above)
Comment (justinemmanuelmercado) [reply test/unit/__snapshots__/finances.test.ts.snap:1653]: I think this is correct

`&quot;` = `"` 

so it's backslash to negate the following backslash to negate the `"`

File: test/unit/__snapshots__/merchant-fulfillment.test.ts.snap [hunk 3]
```diff
@@ -322,7 +322,7 @@ Array [
             "SellerInputDefinition": Object {
               "Constraints": Array [
                 Object {
-                  "ValidationRegEx": "^[a-zA-Z0-9-._&amp; ]{3,25}$",
+                  "ValidationRegEx": "^[a-zA-Z0-9-._& ]{3,25}$",
```
Comment (justinemmanuelmercado) [inline test/unit/__snapshots__/merchant-fulfillment.test.ts.snap:325]: This one get's parsed correctly too

`&amp;` = `&`

File: test/unit/__snapshots__/finances.test.ts.snap (hunk 1 above)
Comment (justinemmanuelmercado) [reply test/unit/__snapshots__/finances.test.ts.snap:1653]: So to me it looks like it's replacing the HTML entities to their `string` counterpart

**Comment by justinemmanuelmercado** [comment]
> On the other hand fast XML parser is TypeScript native.

That's a good point. I could've sworn `xml2js` had its own `.d.ts` file, but apparently something installed `@types/xml2js` because I don't remember installing it

**Comment by justinemmanuelmercado** [comment]
Can't get `xml2json` to work without substantial changes to both tests and codecs.

It's mostly because of how it parses empty or self closing tags, i.e. `<Condition/>`, to `{}` instead of `''` like how `fast-xml-parser` or `xml2js` parses empty or self closing tags. See the following https://github.com/ScaleLeap/amazon-mws-api-sdk/blob/de12f95866581d34fa9e9cd485a64fd2a39a3182/test/unit/__snapshots__/fulfillment-inventory.test.ts.snap#L83

This isn't bad in itself but it means changing things like the parsing error unit tests. It throws `ParsingError: Problem with property "ListInventorySupplyByNextTokenResponse": it does not exist in received object {}` instead of `ParsingError: Expected an object, but received null`.

File: test/unit/__snapshots__/merchant-fulfillment.test.ts.snap (hunk 3 above)
Comment (moltar) [reply test/unit/__snapshots__/merchant-fulfillment.test.ts.snap:325]: Yeah, you are totally right! I overlooked that it was actually just escaping the literal `"` char. I thought it was the end of the string, and that `&quot;` was converted to `\` :D Good eye!

So does that mean that fast parser gets it wrong? Shall we open an issue in their repo about it?

**Comment by moltar** [comment]
> Can't get `xml2json` to work without substantial changes to both tests and codecs.

Given that the difference in speed seems to be negligible and fast parser is TS native, I think let's just keep it as is for now.

**Comment by justinemmanuelmercado** [comment]
> Given that the difference in speed seems to be negligible and fast parser is TS native, I think let's just keep it as is for now.

I'm okay with this for now, but I think this might be good to think about in the future. Issues could arise with things like not having a way to access a tags' attributes, and maybe there would be a preference to have the actual character `\\"` in the response as opposed to having the user sanitize the string themselves

**Comment by moltar** [comment]
> I'm okay with this for now, but I think this might be good to think about in the future.

Well, let's keep the branch alive for now. Maybe we'll do it eventually.

> Issues could arise with things like not having a way to access a tags' attributes, and maybe there would be a preference to have the actual character `\\"` in the response as opposed to having the user sanitize the string themselves

These really sound like defects with the library though. Can we maybe contribute a fix to that? Or at least raise an issue in their repo?

**Comment by justinemmanuelmercado** [comment]
> These really sound like defects with the library though. Can we maybe contribute a fix to that? Or at least raise an issue in their repo?

I'm making an issue to resolve the things that became apparent in this PR. I wasn't really fair with `fast-xml-parser`. I think if I tweak the options for FXP's parser, I could fix the attributes and the sanitized strings. Good thing `xml2js` handled attributes differently, that put these types of things into light

**Comment by justinemmanuelmercado** [comment]
Also I'm closing this PR, I think you're right that it'd be better to stick `fast-xml-parser` due to native TS and also `fast-xml-parser` actually has benchmarks of it against `xml2js` in its GH page

**Pull request closed without merging**